src/devices/Scd4x/Scd4x.cs

```diff
@@ -0,0 +1,240 @@
+        public bool CheckDataReady()
+        {
+            TimeSpan delay = BeginCheckDataReady();
+            Thread.Sleep(delay);
+            return EndCheckDataReady();
+        }
+
+        /// <summary>
+        /// <para>
+        /// Begins checking if a periodic measurement is available.
+        /// Once available, the measurement can be read via <see cref="ReadPeriodicMeasurement"/>.
+        /// </para>
+        ///
+        /// <para>
+        /// <see cref="StartPeriodicMeasurements"/> must be called first.
+        /// </para>
+        /// </summary>
+        /// <returns>
+        /// The time that the caller must wait before calling <see cref="EndCheckDataReady"/> to retreive the current data ready status.
+        /// </returns>
+        public TimeSpan BeginCheckDataReady()
+        {
+            _device.Write(CheckDataReadyStatusBytes);
+            return TimeSpan.FromTicks(TimeSpan.TicksPerMillisecond);
+        }
+
+        /// <summary>
+        /// Completes checking if a periodic measurement is available.
+        /// <see cref="BeginCheckDataReady"/> must have been called first.
+        /// </summary>
+        /// <returns>If a measurement is available, <see langword="true"/>. Otherwise, <see langword="false"/>.</returns>
+        public bool EndCheckDataReady()
+        {
+            Span<byte> buffer = stackalloc byte[3];
+            _device.Read(buffer);
+
+            return Sensirion.ReadUInt16BigEndianAndCRC8(buffer) is ushort response && (response & 0x7FF) != 0;
+        }
+
+        /// <summary>
+        /// <para>
+        /// Reads a periodic CO₂, humidity, and temperature measurement from the sensor.
+        /// </para>
+        ///
+        /// <para>
+        /// <see cref="CheckDataReady"/> should be called first to ensure a measurement is available. Once read, data will no longer be ready.
+        /// </para>
+        /// </summary>
+        /// <returns>A tuple of CO₂, humidity, and temperature.</returns>
+        public (VolumeConcentration? CarbonDioxide, RelativeHumidity? RelativeHumidity, Temperature? Temperature) ReadPeriodicMeasurement()
+        {
+            TimeSpan delay = BeginReadPeriodicMeasurement();
+            Thread.Sleep(delay);
+            return EndReadPeriodicMeasurement();
+        }
+
+        /// <summary>
+        /// <para>
+        /// Begins reading a periodic CO₂, humidity, and temperature measurement from the sensor.
+        /// </para>
+        ///
+        /// <para>
+        /// <see cref="CheckDataReady"/> should be called first to ensure a measurement is available. Once read, data will no longer be ready.
+        /// </para>
+        /// </summary>
+        /// <returns>
+        /// The time that the caller must wait before calling <see cref="EndReadPeriodicMeasurement"/> to retreive the measurement.
+        /// </returns>
+        public TimeSpan BeginReadPeriodicMeasurement()
```

Comment:
Same coment here, not sure why we have Begin/End members public. If there is a scenario for running this async, then perhaps we should add Task-based (async and await) methods instead.

Reply:
My goal is to allow a single thread to efficiently interleave talking to multiple devices at a time. Basically:

1. Write command to Device A.
2. Write command to Device B.
3. Wait enough time for both commands to complete.
4. Read response from Device A.
5. Read response from Device B.

There exist similar async APIs in some Arduino libraries.

Tasks feel problematic here:

- Platform APIs for I2C are all sync.
    - This means that async would be "fake" and you'd be reliant on the thread pool scaling up.
- Because I2C only supports a single I/O (steps 1, 2, 4, and 5 above) at a time, these platform APIs do their own locks.
    - This means if you have N outstanding I/Os, you'll have N blocked threads. :(

Admittedly this is probably not a huge issue -- I suspect in many cases one will have very few I2C devices. Maybe it's fine to just ignore the async-over-sync problem and hope nobody loads up their I2C bus. We could improve the scenario a little bit by making I2cDevice have async APIs that synchronize around a bus-wide `SemaphoreSlim`, so only one thread is ever blocked at once. The overhead tradeoffs would be interesting.

I didn't feel it was appropriate to make this choice for the caller, so instead chose to just expose the low-level primitives needed to implement it whatever way one might want, without forcing down an async/await path.

Reply:
I've taken a look at some async code from other drivers and see semi-async drivers like this is a thing. I've updated the code with this.